python/bindings/novatel/py_decoded_message.hpp

```diff
@@ -0,0 +1,45 @@
+#pragma once
+
+#include "bindings_core.hpp"
+#include "novatel_edie/decoders/oem/message_decoder.hpp"
+
+namespace nb = nanobind;
+using namespace novatel::edie;
+
+struct PyGpsTime
+{
+    PyGpsTime() = default;
+    PyGpsTime(uint16_t week_, double milliseconds_, TIME_STATUS time_status_) : week(week_), milliseconds(milliseconds_), time_status(time_status_) {}
+    PyGpsTime(const oem::MetaDataStruct& meta_) : week(meta_.usWeek), milliseconds(meta_.dMilliseconds), time_status(meta_.eTimeStatus) {}
+
+    uint16_t week{0};
+    double milliseconds{0.0};
+    TIME_STATUS time_status{TIME_STATUS::UNKNOWN};
+};
+
+struct PyDecodedMessage
+{
+    explicit PyDecodedMessage(std::vector<FieldContainer> message_, const oem::MetaDataStruct& meta_);
+    nb::dict& get_values() const;
+    nb::dict& get_fields() const;
+    nb::dict to_dict() const;
+    nb::object getattr(nb::str field_name) const;
+    nb::object getitem(nb::str field_name) const;
+    bool contains(nb::str field_name) const;
+    size_t len() const;
+    std::string repr() const;
+
+    std::vector<FieldContainer> fields;
+
+    // MetaDataStruct with all fields that are no longer relevant after decoding dropped.
+    uint16_t message_id;
+    uint32_t message_crc;
+    std::string message_name;
+    PyGpsTime time;
```

Comment:
Most of the details in the MetaData object, besides the ones included here, are of no use to the user after the message has been parsed, really, and passing around a pair of objects in the consumer code seems unnecessarily clumsy to me. I originally only wanted to keep track of the message ID / name for a clearer repr(), but it made sense to copy all relevant fields instead.

I would suggest only returning the PyDecodedMessage from Parser and FileParser, which provide a more abstract interface to the decoding process anyway, and would only return a pair of PyDecodedMessage and MetaData from the more advanced Decoder class.

What do you think?

Also, all of the message names should easily fit within the small string optimization limit of 23 characters, so they are stored in-place rather than on the heap for `message_name`, which is nice. I have not done any serious benchmarking, but I would expect the performance impact of copying over these fields to be negligible and the UI/UX improvement worth the cost.

Reply:
My initial thoughts:

I agree that the metadata is mostly useless after decoding and filtering except for the message name. Embedding that in the header or or body would be ideal. I checked internally and I can see message names up to a maximum of 31 characters so if we could make it that length (or a few extra bytes for safety) is very reasonable idea. I do think that the header should be passed around with the message data as there are a number of fields that are useful (id, port, idle, time status, time, and receiver status are all commonly used either in development or testing) and I think embedding all of them into the message would be a bit too much. Does it make sense to pass around a message object with a Header and Body?

For Parser and FileParser I think we still need the ability to return the header also as it provides critical fields that we need for our testing. I've tried pushing the more advanced framer->headerdecoder->filter->message decoder internally and it didn't get much traction. People much preferred the simpler interface of the Parser and FileParser. I think this behavior would extend to people/companies using this package to so I think providing the Header back, in addition to the body, would be required. We can discuss this further if you disagree.

I agree that the impact of storing the name with the message data (I prefer the header!) would be worth the cost. If we drop the metadata at some point in the process then the Python code gets much worse.

`header.name == 'bestpos'` looks much better than `header.id == 42'`